torch/csrc/jit/script/compiler.cpp

```diff
@@ -474,6 +489,87 @@ struct to_ir {
     }
   }
 
+  void emitFor(const For& stmt) {
+    // For now, we only support range loops. e.g. for i in range(3): ...
+
+    auto targets = stmt.targets();
+    auto itrs = stmt.itrs();
+    auto body = stmt.body();
+
+    // itrs must consist of a single Apply node
+    if (stmt.itrs().size() != 1) {
+      throw ErrorReport(stmt)
+          << "List of iterables is not supported currently.";
+    }
+    if (itrs[0].kind() != TK_APPLY) {
+      throw ErrorReport(stmt)
+          << "Non-range for loops are currently not supported.";
+    }
+
+    Apply range_iterator = Apply(itrs[0]);
+    if (range_iterator.callee().kind() != TK_VAR) {
+      throw ErrorReport(stmt)
+          << "Non-range for loops are currently not supported.";
+    }
+
+    {
+      Var var = Var(range_iterator.callee());
+      if (var.name().name() != "range") {
+        throw ErrorReport(stmt)
+            << "Non-range for loops are currently not supported.";
+      }
+    }
+
+    int max_trip_count_val;
+    {
+      List<Expr> args = range_iterator.inputs();
+      // TODO: start, stop, step loop
+      if (args.size() != 1) {
+        throw ErrorReport(stmt)
+            << "range() expects one argument but got" << args.size();
+      }
+
+      if (args[0].kind() != TK_CONST) {
+        throw ErrorReport(stmt) << "Argument to range() must be a constant";
+      }
+      auto const_tree = args[0].get();
+      auto num_value = const_tree->tree(0)->doubleValue();
+      auto type = const_tree->tree(1)->stringValue();
+      if (type != "i") {
+        throw ErrorReport(stmt) << "Argument to range() must be an integer";
+      }
+      max_trip_count_val = num_value;
+    }
+
+    auto target_list = List<Ident>(targets);
+    std::vector<std::string> target_idents;
+    target_idents.reserve(target_list.size());
+    for (const auto& t : target_list) {
+      target_idents.push_back(Ident(t).name());
+    }
+
+    Value* max_trip_count = emitConst(stmt.range(), max_trip_count_val, "i")[0];
+    Value* cond_value_dummy = emitConst(stmt.range(), 1, "b")[0];
+
+    emitLoopCommon(
+        max_trip_count,
+        cond_value_dummy,
+        stmt.body(),
+        stmt,
```

Comment:
Can we avoid passing `stmt.body` and `stmt`? It's redundant

Reply:
I don't think so, not unless we make a common interface for `struct While` and `struct For` that exposes a virtual `body()`